core/src/test/java/org/apache/iceberg/avro/AvroDataUnionRecordTest.java

```diff
@@ -0,0 +1,184 @@
+
+  protected void writeAndValidate(
+      List<GenericData.Record> expected,
+      List<GenericData.Record> expectedRead,
+      Schema icebergSchema) throws IOException {
+    File testFile = temp.newFile();
+    Assert.assertTrue("Delete should succeed", testFile.delete());
+
+    try (FileAppender<GenericData.Record> writer = Avro.write(Files.localOutput(testFile))
+        .schema(icebergSchema)
+        .named("test")
+        .build()) {
+      for (GenericData.Record rec : expected) {
+        writer.add(rec);
+      }
+    }
+
+    List<GenericData.Record> rows;
+    try (AvroIterable<GenericData.Record> reader = Avro.read(Files.localInput(testFile))
+        .project(icebergSchema)
+        .build()) {
+      rows = Lists.newArrayList(reader);
+    }
+
+    for (int i = 0; i < expectedRead.size(); i += 1) {
+      AvroTestHelpers.assertEquals(icebergSchema.asStruct(), expectedRead.get(i), rows.get(i));
+    }
+  }
+
+  @Test
+  public void testMapOfUnionValues() throws IOException {
+    String schema1 = "{\n" +
+        "  \"name\": \"MapOfUnion\",\n" +
+        "  \"type\": \"record\",\n" +
+        "  \"fields\": [\n" +
+        "    {\n" +
+        "      \"name\": \"map\",\n" +
+        "      \"type\": [\n" +
+        "        \"null\",\n" +
+        "        {\n" +
+        "          \"type\": \"map\",\n" +
+        "          \"values\": [\n" +
+        "            \"null\",\n" +
+        "            \"boolean\",\n" +
+        "            \"int\",\n" +
+        "            \"long\",\n" +
+        "            \"float\",\n" +
+        "            \"double\",\n" +
+        "            \"bytes\",\n" +
+        "            \"string\"\n" +
+        "          ]\n" +
+        "        }\n" +
+        "      ]\n" +
+        "    }\n" +
+        "  ]\n" +
+        "}";
+    org.apache.avro.Schema avroSchema = new org.apache.avro.Schema.Parser().parse(schema1);
+    org.apache.iceberg.Schema icebergSchema = AvroSchemaUtil.toIceberg(avroSchema);
+    org.apache.avro.Schema avroSchemaUnionRecord = AvroSchemaUtil.convert(icebergSchema, "test");
+    org.apache.avro.Schema unionRecordSchema =
+        avroSchemaUnionRecord.getFields().get(0).schema().getTypes().get(1).getValueType().getTypes().get(1);
+
+    List<GenericData.Record> expectedRead = new ArrayList<>();
+    List<GenericData.Record> expected = new ArrayList<>();
+
+    for (long i = 0; i < 10; i++) {
+      Map<String, Object> map = new HashMap<>();
+      Map<String, Object> mapRead = new HashMap<>();
+      updateMapsForUnionSchema(unionRecordSchema, map, mapRead, i);
+      GenericData.Record recordRead = new GenericRecordBuilder(avroSchema)
+          .set("map", mapRead)
+          .build();
+      GenericData.Record record = new GenericRecordBuilder(avroSchema)
+          .set("map", map)
+          .build();
+      expected.add(record);
+      expectedRead.add(recordRead);
+    }
+    writeAndValidate(expected, expectedRead, icebergSchema);
```

Comment:
Shouldn't be `actualWrite` and `expectedRead` ?

Reply:
ack